File: src/Sekiban.Pure.Dapr/Parts/PartitionKeysAndProjector.cs
using ResultBoxes;
using Sekiban.Pure.Aggregates;
using Sekiban.Pure.Documents;
using Sekiban.Pure.Projectors;

namespace Sekiban.Pure.Dapr.Parts;

/// <summary>
/// Holds partition keys and projector information for aggregate actors.
/// This is the Dapr equivalent of Orleans' PartitionKeysAndProjector.
/// </summary>
public class PartitionKeysAndProjector
{
    public PartitionKeys PartitionKeys { get; init; }
    public IAggregateProjector Projector { get; init; }

    public PartitionKeysAndProjector(PartitionKeys partitionKeys, IAggregateProjector projector)
    {
        PartitionKeys = partitionKeys ?? throw new ArgumentNullException(nameof(partitionKeys));
        Projector = projector ?? throw new ArgumentNullException(nameof(projector));
    }

    /// <summary>
    /// Creates PartitionKeysAndProjector from a grain key string
    /// </summary>
    public static ResultBox<PartitionKeysAndProjector> FromGrainKey(
        string grainKey,
        IAggregateProjectorSpecifier projectorSpecifier)
    {
        try
        {
            // Extract projector type and partition keys from the grain key
            // Format: "ProjectorType:PartitionKeysString"
            var parts = grainKey.Split(':', 2);
            if (parts.Length != 2)
            {
                return ResultBox<PartitionKeysAndProjector>.FromException(
                    new ArgumentException($"Invalid grain key format: {grainKey}"));
            }

            var projectorTypeName = parts[0];
            var partitionKeysString = parts[1];

            // Get projector from specifier
            var projectorResult = projectorSpecifier.GetProjector(projectorTypeName);
            if (!projectorResult.IsSuccess)
            {
                return ResultBox<PartitionKeysAndProjector>.FromException(
                    projectorResult.GetException());
            }

            var projector = projectorResult.UnwrapBox();

            // Parse partition keys
            var partitionKeysResult = PartitionKeys.FromPrimaryKeysString(partitionKeysString);
            if (!partitionKeysResult.IsSuccess)
            {
                return ResultBox<PartitionKeysAndProjector>.FromException(
                    partitionKeysResult.GetException());
            }

            return ResultBox<PartitionKeysAndProjector>.FromValue(
                new PartitionKeysAndProjector(partitionKeysResult.UnwrapBox(), projector));
        }
        catch (Exception ex)
        {
            return ResultBox<PartitionKeysAndProjector>.FromException(ex);
        }
    }

    /// <summary>
    /// Converts to a grain key string for the projector
    /// </summary>
    public string ToProjectorGrainKey()
    {
        return $"{Projector.GetType().Name}:{PartitionKeys.ToPrimaryKeysString()}";
    }

    /// <summary>
    /// Converts to a grain key string for the event handler
    /// </summary>
    public string ToEventHandlerGrainKey()
    {
        return PartitionKeys.ToPrimaryKeysString();
    }
}

File: ts/src/packages/dapr/src/parts/partition-keys-and-projector.ts
import { PartitionKeys, IAggregateProjector } from '@sekiban/core';
import { Result, ok, err } from 'neverthrow';

/**
 * Holds partition keys and projector information for aggregate actors.
 * This is the TypeScript equivalent of C# Dapr's PartitionKeysAndProjector.
 */
export class PartitionKeysAndProjector<TPayload> {
  constructor(
    public readonly partitionKeys: PartitionKeys,
    public readonly projector: IAggregateProjector<TPayload>
  ) {
    if (!partitionKeys) {
      throw new Error('partitionKeys cannot be null');
    }
    if (!projector) {
      throw new Error('projector cannot be null');
    }
  }

  /**
   * Creates PartitionKeysAndProjector from a grain key string
   * Format: "PartitionKeysString=ProjectorType" (matching C# format)
   */
  static fromGrainKey<TPayload>(
    grainKey: string,
    projectorRegistry: Map<string, IAggregateProjector<TPayload>>
  ): Result<PartitionKeysAndProjector<TPayload>, Error> {
    try {
      // Extract projector type and partition keys from the grain key
      const parts = grainKey.split('=');
      if (parts.length !== 2) {
        return err(new Error(`Invalid grain key format: ${grainKey}`));
      }

      const partitionKeysString = parts[0];
      const projectorTypeName = parts[1];

      // Get projector from registry
      const projector = projectorRegistry.get(projectorTypeName);
      if (!projector) {
        return err(new Error(`Projector not found: ${projectorTypeName}`));
      }

      // Parse partition keys
      const partitionKeys = PartitionKeys.fromPrimaryKeysString(partitionKeysString);

      return ok(new PartitionKeysAndProjector(partitionKeys, projector));
    } catch (error) {
      return err(error instanceof Error ? error : new Error('Unknown error'));
    }
  }

  /**
   * Converts to a grain key string for the projector
   * Matches C# ToProjectorGrainKey()
   */
  toProjectorGrainKey(): string {
    return `${this.partitionKeys.toPrimaryKeysString()}=${this.projector.constructor.name}`;
  }

  /**
   * Converts to a grain key string for the event handler
   * Matches C# ToEventHandlerGrainKey()
   */
  toEventHandlerGrainKey(): string {
    return this.partitionKeys.toPrimaryKeysString();
  }
}